Type conversion (maybe corrupting data)?

Hi 

I am trying to update my R package sundialr which is an interface to some of the solvers in SUNDIALS. The CRAN checks for my package give -Woverflow warnings. I discussed this issue in r-package-devel forum where another user thought I should bring it to the attention of the library since it may be corrupting data.

See the entire issue here - 

https://stat.ethz.ch/pipermail/r-package-devel/2024q1/010431.html

I am not able to submit to CRAN due to this error.

Specifically, these are the warnings I am getting

>Flavor: r-devel-windows-x86_64
>Check: whether package can be installed, Result: WARNING
>  Found the following significant warnings:
>    ./sundials/sundials/sundials_hashmap.h:26:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'14695981039346656037' to '2216829733' [-Woverflow]
>    ./sundials/sundials/sundials_hashmap.h:27:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'1099511628211' to '435' [-Woverflow]
>    sundials/sundials/sundials_hashmap.h:26:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'14695981039346656037' to '2216829733' [-Woverflow]
>    sundials/sundials/sundials_hashmap.h:27:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'1099511628211' to '435' [-Woverflow]
>    sundials/sundials/sundials_profiler.c:71:24: warning: function
>declaration isn't a prototype [-Wstrict-prototypes]
>  See 'd:/RCompile/CRANincoming/R-devel/sundialr.Rcheck/00install.out' for
>details.
>  Used C++ compiler: 'g++.exe (GCC) 12.3.0'

and 

>Flavor: r-devel-linux-x86_64-debian-gcc
>Check: whether package can be installed, Result: WARNING
>  Found the following significant warnings:
>    sundials/sundials/sundials_profiler.c:71:41: warning: a function
>declaration without a prototype is deprecated in all versions of C
>[-Wstrict-prototypes]
>  See '/srv/hornik/tmp/CRAN/sundialr.Rcheck/00install.out' for details.
>  Used C++ compiler: 'Debian clang version 17.0.6 (5)'


Any update on this would be highly appreciated!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -157,6 +157,10 @@ sundials_band.h
 
 Fixed [#329](https://github.com/LLNL/sundials/issues/329) so that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+[#409](https://github.com/LLNL/sundials/issues/409) and
+[#249](https://github.com/LLNL/sundials/issues/249)
+
 ## Changes to SUNDIALS in release 6.7.0
 
 Added the `SUNAdaptController` base class, ported ARKODE's internal

--- a/doc/arkode/guide/source/Introduction.rst
+++ b/doc/arkode/guide/source/Introduction.rst
@@ -314,7 +314,7 @@ Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
 Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
-`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
 `#249 <https://github.com/LLNL/sundials/issues/249>`_.
 
 Changes in v5.7.0

--- a/doc/arkode/guide/source/Introduction.rst
+++ b/doc/arkode/guide/source/Introduction.rst
@@ -313,6 +313,10 @@ rely on these are recommended to transition to the corresponding :c:type:`SUNMat
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`#249 <https://github.com/LLNL/sundials/issues/249>`_.
+
 Changes in v5.7.0
 -----------------
 

--- a/doc/cvode/guide/source/Introduction.rst
+++ b/doc/cvode/guide/source/Introduction.rst
@@ -283,6 +283,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`#249 <https://github.com/LLNL/sundials/issues/249>`_.
+
 Changes in v6.7.0
 -----------------
 

--- a/doc/cvode/guide/source/Introduction.rst
+++ b/doc/cvode/guide/source/Introduction.rst
@@ -284,7 +284,7 @@ Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
 Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
-`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
 `#249 <https://github.com/LLNL/sundials/issues/249>`_.
 
 Changes in v6.7.0

--- a/doc/cvodes/guide/source/Introduction.rst
+++ b/doc/cvodes/guide/source/Introduction.rst
@@ -283,6 +283,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`#249 <https://github.com/LLNL/sundials/issues/249>`_.
+
 Changes in v6.7.0
 -----------------
 

--- a/doc/cvodes/guide/source/Introduction.rst
+++ b/doc/cvodes/guide/source/Introduction.rst
@@ -284,7 +284,7 @@ Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
 Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
-`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
 `#249 <https://github.com/LLNL/sundials/issues/249>`_.
 
 Changes in v6.7.0

--- a/doc/ida/guide/source/Introduction.rst
+++ b/doc/ida/guide/source/Introduction.rst
@@ -245,7 +245,7 @@ Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
 Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
-`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
 `#249 <https://github.com/LLNL/sundials/issues/249>`_.
 
 Changes in v6.7.0

--- a/doc/ida/guide/source/Introduction.rst
+++ b/doc/ida/guide/source/Introduction.rst
@@ -244,6 +244,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`#249 <https://github.com/LLNL/sundials/issues/249>`_.
+
 Changes in v6.7.0
 -----------------
 

--- a/doc/idas/guide/source/Introduction.rst
+++ b/doc/idas/guide/source/Introduction.rst
@@ -258,6 +258,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`#249 <https://github.com/LLNL/sundials/issues/249>`_.
+
 Changes in v5.7.0
 -----------------
 

--- a/doc/idas/guide/source/Introduction.rst
+++ b/doc/idas/guide/source/Introduction.rst
@@ -259,7 +259,7 @@ Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
 Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
-`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
 `#249 <https://github.com/LLNL/sundials/issues/249>`_.
 
 Changes in v5.7.0

--- a/doc/kinsol/guide/source/Introduction.rst
+++ b/doc/kinsol/guide/source/Introduction.rst
@@ -239,6 +239,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`#249 <https://github.com/LLNL/sundials/issues/249>`_.
+
 Changes in v6.7.0
 -----------------
 

--- a/doc/kinsol/guide/source/Introduction.rst
+++ b/doc/kinsol/guide/source/Introduction.rst
@@ -240,7 +240,7 @@ Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
 Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
-`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
+`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
 `#249 <https://github.com/LLNL/sundials/issues/249>`_.
 
 Changes in v6.7.0

--- a/src/sundials/sundials_hashmap.h
+++ b/src/sundials/sundials_hashmap.h
@@ -20,14 +20,15 @@
 #ifndef _SUNDIALS_HASHMAP_H
 #define _SUNDIALS_HASHMAP_H
 
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "sundials/sundials_errors.h"
 #include "sundials/sundials_types.h"
 
-static const unsigned long HASH_PRIME        = 14695981039346656037U;
-static const unsigned long HASH_OFFSET_BASIS = 1099511628211U;
+static const uint64_t HASH_PRIME        = 14695981039346656037U;
+static const uint64_t HASH_OFFSET_BASIS = 1099511628211U;
 
 /*
   For a nice discussion on popular hashing algorithms see:
@@ -36,9 +37,9 @@ static const unsigned long HASH_OFFSET_BASIS = 1099511628211U;
   This is a 64-bit implementation of the 'a' modification of the
   Fowler–Noll–Vo hash (i.e., FNV1-a).
  */
-static unsigned long fnv1a_hash(const char* str)
+static uint64_t fnv1a_hash(const char* str)
 {
-  unsigned long hash = HASH_OFFSET_BASIS;
+  uint64_t hash = HASH_OFFSET_BASIS;
   char c;
   while ((c = *str++)) { hash = (hash ^ c) * HASH_PRIME; }
   return hash;

